core/src/packet.rs

```diff
@@ -124,21 +127,61 @@ impl Meta {
 
 #[derive(Debug, Clone)]
 pub struct Packets {
-    pub packets: Vec<Packet>,
+    pub packets: PinnedVec<Packet>,
+
+    recycler: Option<PacketsRecycler>,
+}
+
+impl Drop for Packets {
+    fn drop(&mut self) {
+        if let Some(ref recycler) = self.recycler {
+            let old = mem::replace(&mut self.packets, PinnedVec::default());
+            recycler.recycle(old)
+        }
+    }
+}
+
+impl Reset for Packets {
+    fn reset(&mut self) {
+        self.packets.resize(0, Packet::default());
```

Comment:
this implementation of Reset looks to conflict with the implementation of Default...

recycler uses default() to construct, then immediately throws away the allocation?

think reset() should truncate(NUM_RCVMMSGS) instead?

you've also touched the streamer code, removing the special-ness of filling NUM_RCVMMSGS per call to recv_mmsg(), though, right?

Reply:
> this implementation of Reset looks to conflict with the implementation of Default...

They both give you back a size=0 vec. The reserve size is just an optimization, it doesn't affect the functionality of the vec, just how many times it would be realloc'ed if resized.

> think reset() should truncate(NUM_RCVMMSGS) instead?

No, size should be 0 for construction, capacity could be NUM_RCVMMSGS, but doesn't have to be. I actually use a special reserve of 256 packets for the streamer paths because resize by NUM_RCVMMSGS at a time causes quite a few allocations, about 16 allocations per vector to get up to about the ~240-250 packets length. This will be the reserve for whatever cases we use Packets in like when we create dummy packets to send in a test.

>you've also touched the streamer code, removing the special-ness of filling NUM_RCVMMSGS per call to recv_mmsg(), though, right?

Well, it still tries to fill NUM_RCVMMSGS at a time, but it doesn't necessarily require that many to stop filling the Packets buffer.